File: TP1/comeandgo.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm> 
using namespace std;

void add_edge(int i, int j, int type, vector<set<int>>& graph) {
    if (type == 1) {
        graph[i].insert(j);
    } else if (type == 2) {
        graph[i].insert(j);
        graph[j].insert(i);
    }
}

void DFS_visited(int u, vector<bool>& visited, vector<set<int>>& graph) {
    visited[u] = true;
    for (auto v : graph[u]) {
        if (!visited[v]) {
            DFS_visited(v, visited, graph);
        }
    }
}

int DFS(vector<set<int>>& graph) {
    for (int start = 0; start < graph.size(); ++start) {
        vector<bool> visited(graph.size(), false);
        DFS_visited(start, visited, graph);
        if (std::any_of(visited.begin(), visited.end(), [](bool v){ return !v; })) {
            return 0;
        }
    }
    return 1;
}
int main() {
    int vertices, edges;
    while (cin >> vertices >> edges) {
        if (edges == 0) break;
        vector<set<int>> graph(vertices+1);
        for (int i = 0; i < edges; ++i) {
            int source, target, type;
            cin >> source >> target >> type;
            add_edge(source, target, type, graph);
            if (i+1 == edges) {
                cout << DFS(graph) << endl;
            }
        }
    }
    return 0;
}

File: TP1/comeandgo.py
def build_matrix(n:int, value = 0):
    matrix = [[value for _ in range(n)] for _ in range(n)]
    return matrix

def add_edge(i, j, type, graph):
    graph[i][j] = 1
    if type ==2:
        graph[j][i] = 1
    return graph

def DFS_visited(graph, u, visited):
    visited[u] = True
    for v in range(len(graph)):
        if graph[u][v] == 1 and not visited[v]:
            DFS_visited(graph, v, visited)
    return visited

def DFS(graph):
    for start in range(len(graph)):
        visited = [False for _ in range(len(graph))]
        visited = DFS_visited(graph, start, visited)
        if False in visited:  
            return 0
    return 1  

while True:
    vertices, edges = map(int, input().split())
    if edges == 0 or vertices ==0: break
    graph = build_matrix(vertices)
    for i in range(edges):
        source, target, type = map(int, input().split()) 
        add_edge(source-1, target-1, type, graph)
        
        
    print(DFS(graph))
        